server/src/game/player_message.rs

```diff
@@ -0,0 +1,85 @@
+use uuid::Uuid;
+use actix::{Handler, Message, Addr, Context, AsyncContext};
+
+use crate::engine::{WebSocket, websocket_message};
+use crate::game::Player;
+use chrono::{DateTime, Utc};
+
+#[derive(Message)]
+#[rtype(result = "()")]
+pub struct Connected {
+    pub websocket_addr: Addr<WebSocket>,
+    pub session_id: Uuid,
+}
+
+impl Handler<Connected> for Player {
+    type Result = ();
+
+    fn handle(&mut self, msg: Connected, ctx: &mut Context<Self>) -> Self::Result {
+        if let None = self.websocket_addr {
+            self.websocket_addr = Some(msg.websocket_addr.clone());
+            msg.websocket_addr.do_send(websocket_message::UserConnected {
+                session_id: msg.session_id,
+                username: self.username.clone(),
+                player_addr: ctx.address(),
+            });
+        } else {
+            msg.websocket_addr.do_send(websocket_message::ConnectionFailed {
+                reason: "User still connected.".to_string()
+            });
+        }
+    }
+}
+
+#[derive(Message)]
+#[rtype(result = "()")]
+pub struct Disconnected {}
+
+impl Handler<Disconnected> for Player {
+    type Result = ();
+
+    fn handle(&mut self, _msg: Disconnected, _ctx: &mut Context<Self>) -> Self::Result {
+        self.websocket_addr = None;
+        // TODO: wait, then skip in game, later drop from game
+    }
+}
+
+#[derive(Message)]
+#[rtype(result = "()")]
+pub struct GlobalChat {
+    pub message: String,
+}
+
+impl Handler<GlobalChat> for Player {
+    type Result = ();
+
+    fn handle(&mut self, msg: GlobalChat, _ctx: &mut Context<Self>) -> Self::Result {
+        self.server_addr.do_send(GlobalChatBroadcast {
+            timestamp: Utc::now(),
+            username: self.username.clone(),
+            message: msg.message,
+        })
+    }
+}
```

Comment:
i think it would be great if you had some naming convention to distinguish message types from each other, this part of the code shows it pretty well:
`struct GlobalChat` and `struct GlobalChatBroadcast` from the name, no difference. First takes a message from the player and sends it to the server, the second, it sends a message to the player client, from the server.

Theres a lot of messages here that could be maybe renamed to commands, either from the viewpoint of the server, or the viewpoint of the client, but it would be great if it was the same everywhere, its very hard to know what the message is doing from the name + package, you have to go into the handler impl to find out what it does. So like maybe Sent/Received<insert something>, or GlobalChatOut/In, or <message>Server/Client, just to make it extremely obvious what direction the message is going in, whether from server -> client or from client -> server.